### PR TITLE
Fix sliding edge-case in Carousel (#3654)

### PR DESCRIPTION
**Pull request by bancho22**

### CODE REVIEW

**Comment by bancho22** [comment]
I will need to do a better job with the test here. Any good suggestions?

File: test/CarouselSpec.js [hunk 1]
```diff
@@ -238,4 +238,13 @@ describe('<Carousel>', () => {
     const wrapper = mount(<Carousel>{items}</Carousel>);
     wrapper.find('div').length.should.equal(4);
   });
+
+  it('Should move backwards if left control is clicked on the first item', () => {
+    const wrapper = mount(<Carousel defaultActiveIndex={0}>{items}</Carousel>);
+    wrapper.find('a.carousel-control-prev').simulate('click');
+
+    setTimeout(() => {
+      wrapper.assertSingle('.carousel-item-left');
```
Comment (taion) [inline test/CarouselSpec.js]: This assertion isn't actually going to fire, because the test case isn't async.

Perhaps make it async? Or else take a `done` callback in the fn.

Comment (mxschmitt) [reply test/CarouselSpec.js]: You could also use 
````
const clock = sinon.useFakeTimers({
      toFake: ['setTimeout'],
});
// some logic
clock.tick(100);
// check if X has happened
```

File: src/Carousel.js [hunk 2]
```diff
@@ -151,7 +151,8 @@ class Carousel extends React.Component {
       let direction;
       if (
         (nextIndex === 0 && previousActiveIndex >= lastPossibleIndex) ||
-        previousActiveIndex <= nextIndex
+        (previousActiveIndex <= nextIndex &&
+          !(previousActiveIndex === 0 && nextIndex >= lastPossibleIndex))
```
Comment (alvinbabujohn) [inline src/Carousel.js]: Not a bad fix, but still will have issues when children=2, no?

**Comment by mxschmitt** [comment]
@bancho22 what's the state of this PR?

Could you fix the review comments? otherwise we have to close the PR.

File: test/CarouselSpec.js (hunk 1 above)
Comment (bancho22) [reply test/CarouselSpec.js]: Thanks! Turns out I planned the test case all wrong. I don't really see a good way to test this.

File: src/Carousel.js (hunk 2 above)
Comment (bancho22) [reply src/Carousel.js]: If children=2, then lastPossibleIndex will always be equal to 1 and so would nextIndex.
Given that nextIndex is always 1, if previousActiveIndex is indeed zero or one which it would be, then the if condition would always be satisfied.
Thus, you are correct, however, I believe it's in a better condition this way as at least we fix the most common use cases.

**Comment by bancho22** [comment]
@mxschmitt There is a problem with the Travis CI build that I don't understand. Otherwise, one of the comments, I resolved, while the other one seems like it would be unresolvable, unless we agree we can live with that small issue.

**Comment by mxschmitt** [comment]
> @mxschmitt There is a problem with the Travis CI build that I don't understand. Otherwise, one of the comments, I resolved, while the other one seems like it would be unresolvable, unless we agree we can live with that small issue.

The Travis CI issue is fixed.

**Comment by taion** [comment]
This logic is super screwy, BTW. `handleNext` and `handlePrev` actually call `onSelect` with the direction, but the gDSFP logic just ignores that as far as I can tell.

**Comment by taion** [comment]
superseded by https://github.com/react-bootstrap/react-bootstrap/pull/4917
